### DISCUSSION TITLE
When doing `fit()`, `self.training` in `forward()` keeps turning into False?

### DISCUSSION THREAD

**Discussion by HaFred** [Lightning-AI/pytorch-lightning#11838] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hi all, I tried to train a model with pl. And I just ran the `trainer.fit()` as below:
```python
trainer.fit(model, train_dataloaders=model.train_dataloader(),
                val_dataloaders=model.val_dataloader())
```
And I found that `model.training == False`, when it gets into `forward()`...
Is there any solution or does anybody know the potential reason for this?
Or does anyone know where can I find the source code for `training_step()` so that I can check and debug `forward()` in `fit()`?

Thank you very much.

**Comment by rohitgr7** [comment, marked as answer]
does it print `self.training = False` for all the training steps? maybe you might have checked it during the initial steps where val sanity check happens.

**Reply by HaFred** [reply]
Thanks for your reply. After deactivating the val sanity check, `self.training` can remain `True` for `fit()` now. By the way, do you know where to locate the src code for `training_step()`?

The best one I can locate is as in https://github.com/PyTorchLightning/pytorch-lightning/blob/6f22b3623c28028026b3cb8bb534c1ebca9c5ac8/pytorch_lightning/core/lightning.py#L563

However, I am expecting to trace something maybe ATEN or other C++ lib built upon ATEN,  just like PyTorch?

**Reply by rohitgr7** [reply]
You mean the call for this method?

**Reply by HaFred** [reply]
Actually, I meant the definition. Thanks.

**Reply by rohitgr7** [reply]
maybe I didn't get you here... it just calls the method you have defined within your lightningmodule. So the definition is up to the user.

**Reply by HaFred** [reply]
Sorry for the confusion. Yes, I meant the call. I saw `is_param_in_hook_signature()` is calling `training_step()`, but as a further wrapper. It is kinda hard to trace the mechanism of how `training_step()` works as in PyTorch's counterparts?
```python
loss.backward()
optimizer.step()
```

**Reply by rohitgr7** [reply]
well this is the module where everything happens: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/optimization/optimizer_loop.py

training step isn't called explicitly but is called as a part of the closure which is passed to optimizer. So the closure warps call definitions of training_step, backward and optimizer.step and passed to the optimizer
https://github.com/PyTorchLightning/pytorch-lightning/blob/914f685ed826d9adf931b60348cb587e7bec35e5/pytorch_lightning/loops/optimization/optimizer_loop.py#L97-L131
